### DISCUSSION TITLE
Problem with pyznap if you upgrade from older Ubuntu version to 24.04

### DISCUSSION THREAD

**Discussion by vitali-fridman** [Sithuk/ubuntu-server-zfsbootmenu#50] [General]
I upgraded from Ubuntu 23.10 to 24.04 and pyznap stopped working. It's especially problematic because it's called on any apt commands. It appears that root of the problem is that in pyznap virtual environment python is upgraded to 3.12 while the folder with python libraries is still under old version. I solved it by renaming /opt/pyznap/venv/lib/(old version name) to 'python3.12'.

**Comment by vitali-fridman** [comment]
I just run into this problem on another system and renaming lib folder did not work. In this case one of the modules used by pyznap had changes that made it incompatible after the upgrade. I don't know much about python but maybe installing pyznap without virtual environment will solve the problem. Hopefully both python and system level modules will be upgraded with OS upgrade. If that it not the case likely manual intervention is required.

**Reply by Sithuk** [reply]
I updated the pyznap compilation code in the latest version of the script for 24.04 compatibility. Have you tried the latest compilation code?

**Reply by Sithuk** [reply]
A snap of pyznap would solve the dependency issue. However, there are two main issues with using a snap.
1. Snaps cannot be installed in a chroot environment. Although given the increasing prevalence of snaps in ubuntu, the script will likely need a re-write to be snap compatible anyway.
2. There is no pyznap snap available. I wanted to keep the script transparent for review and have therefore used direct compilation from a github source or a source binary where I can. I have avoided creating and publishing a pyznap snap for that reason.

When I got pyznap to compile for 22.04 it solved the issue for now. However, I think the long term direction is to create a snap for pyznap, publish it, and then use it in future releases. There are multiple requests on issue threads on the pyznap github project requesting the author create a flatpak or snap, but with no reply. Pyznap appears to be a dormant project now so I doubt the author will create the snap.

Another alternative is to find a pynap replacement that doesn't wake sleeping drives when taking snapshots. Do you know of one?

**Comment by vitali-fridman** [comment]
Unfortunately I can't test your new code with straight 24.04 install. As I indicated in another thread my systems do not boot when I try it. So I install 22.04 followed by upgrade to 24.04 and that is when pyznap incompatibilities show up. As a workaround I comment out pyznap install from your script and I do it as a separate step after the upgrade. This is of cause is less then ideal. I don't know any good pyznap replacements.

**Reply by Sithuk** [reply]
I mean to try using the pyznap compilation code that is in the latest version of the script after your upgrade. I understand from your response that doing so installs a working pyznap version?

Sanoid and zrepl are popular pyznap alternatives that appear currently maintained.
https://github.com/jimsalterjrs/sanoid
https://github.com/zrepl/zrepl

**Comment by vitali-fridman** [comment]
Yes, I run you pyznapinstall after the upgrade and it works perfectly. If it's going to break again in the future depends on what possible python versioning incompatibilities may be introduced. One suggestion is to make pyznapinstall execution optional as well as to make new script parameter that execute just that part.  After OS upgrade there will be a need to run "sudo pyznap snap" and if it errors out execute your script with that new parameter. My understanding sanoid is written in perl and zepl in go. I have no idea how these languages manage versioning.

**Comment by Sithuk** [comment]
That's good to know, thank you for the feedback. I will consider making the pyznap reinstall an option. I have already done the same with reinstalling zfsbootmenu.
Zrepl maintains a ubuntu repo. Sanoid is available as an ubuntu package in the standard repositories so there should not be dependency issues with those.
https://zrepl.github.io/installation/apt-repos.html
https://packages.ubuntu.com/search?keywords=sanoid

**Comment by Sithuk** [comment]
There is another user asking a question about upgrading on another thread. Could you share your experience on upgrading? It’s not something I’ve tested.

**Reply by timus-22** [reply]
> There is another user asking a question about upgrading on another thread. Could you share your experience on upgrading? It’s not something I’ve tested.

First of all, I want to thank the author for creating such an enchanting script, a very useful tool for novice users who want to get rid of this status as soon as possible! I switched to ZFS myself after using ext4. I will not describe the reasons for the transition, there are quite a lot of advantages of ZFS on the network.  I understand that I have not yet exhausted the full functionality of the script, it is more intended for remote management of snapshots on servers, but the fact that it helped me in some part in installing and using the system with snapshots and rollbacks on a laptop without dancing with a tambourine, for this it has great respect and respect, as we have on the language of birches and aspens is spoken. I experienced these dances by installing Ubuntu with zsys software https://openzfs.github.io/openzfs-docs/Getting%20Started/Ubuntu/Ubuntu%2022.04%20Root%20on%20ZFS.html The path, to put it mildly, is not for the reliability of the system and the data on it, here you can find and express a large stream of unflattering words to Canonical, who considered this option of using ZFS to be the right solution, but I am not the author, and they know better.
The first acquaintance with the script by the author of Sithuk took place in early 2025, according to the script from 2025-01-13, I installed an image on my laptop ubuntu-22.04.5-desktop-amd64.iso with minimal server software, and then installed Mint Cinnamon on top. Subsequent regular updates of Linux Mint with the change of the Ubuntu database passed without hesitation. The current state is Linux Mint 22.2 Zara base: Ubuntu 24.04 noble and I still use pyznap to manage snapshots. I have not observed any problems with creating snapshots on a separate USB SSD in datapool. All scheduled tasks in /etc/cron.d/pyznap were working perfectly. I would not like to experience the moment when pyznap stops performing its functionality after the next system update, but if it does happen, you will have to switch to snapshot management via sanoid.
There are questions about using the system after installation, but they will be indicated to the author and the community in a separate post.
